### DISCUSSION TITLE
Infer properly Error of a Result

### DISCUSSION THREAD

**Discussion by pand-app** [bloodyowl/boxed#82] [Q&A] [answered]
Hello, 

I have three functions : `Child1`, `Child2` and `Parent`.
All of them are returning `Result`.
In `Parent`, I want to type the Result with errors coming from `Child1` and `Child2`. 
I'm able to write intermediary types like `Child1Errors` and `Child2Errors` then combining them in `Parent` like `Result<_, Child1Errors | Child2Errors>`.

It isn't possible to infer errors from a function ? I've tried to do `type InferError<T> = T extends Result<any, infer E> ? E : never;` but it lead to a exhaustive branch malfunction using `ts-pattern`…

Any idea ?

**Comment by zoontek** [comment]
@Nainterceptor Could you create an exemple in the Typescript playground? That's not super clear.

**Reply by pand-app** [reply]
https://stackblitz.com/edit/m21xpf?file=index.ts
-> Here, types are OK

https://stackblitz.com/edit/m21xpf-ket3t3?file=index.ts
-> Here, I've an issue with my type `InferError<T> = T extends Result<any, infer E> ? E : never;`

I think that it's coming from the `never` part, but I don't know how to replace that.

```
error: TS2322 [ERROR]: Type 'Error<null, Child1Error>' is not assignable to type 'Result<null, never>'.
  Type 'Error<null, Child1Error>' is not assignable to type 'Error<null, never>'.
    Type 'Child1Error' is not assignable to type 'never'.
    return result1;
    ~~~~~~
    at file:///Users/·/scratch_6.ts:30:5

TS2322 [ERROR]: Type 'Error<null, Child2Error>' is not assignable to type 'Result<null, never>'.
  Type 'Error<null, Child2Error>' is not assignable to type 'Error<null, never>'.
    Type 'Child2Error' is not assignable to type 'never'.
    return result2;
    ~~~~~~
    at file:///Users/·/scratch_6.ts:35:5

TS2345 [ERROR]: Argument of type '{ readonly tag: "Error"; readonly error: Chainable<GuardP<unknown, Child2Error>>; }' is not assignable to parameter of type 'PatternMatcher<never>'.
  Property '[matcher]' is missing in type '{ readonly tag: "Error"; readonly error: Chainable<GuardP<unknown, Child2Error>>; }' but required in type 'Matcher<never, unknown, any, any, unknown>'.
    Result.P.Error(P.instanceOf(Child2Error)),
    ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    at file:///Users/·/scratch_6.ts:48:5

    '[matcher]' is declared here.
        [matcher](): MatcherProtocol<input, narrowed, matcherType, selections, excluded>;
        ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
        at file:///Users/·/ts-pattern/5.3.1/dist/types/Pattern.d.ts:24:5

Found 3 errors.

```

**Reply by zoontek** [reply, marked as answer]
@Nainterceptor That's normal, because in your example, you have this:

```ts
type InferError<T> = T extends Result<any, infer E> ? E : never;
```

And this:

```ts
function Parent(): Result<
  null,
  InferError<typeof Child1> | InferError<typeof Child2>
>
```

But `typeof Child1` and `typeof Child2` are not of type `Result`, but of type `() => Result<null, Child1Error>` and `() => Result<null, Child2Error>`.

To fix it, you can ever either use `ReturnType`:

```ts
function Parent(): Result<
  null,
  InferError<ReturnType<typeof Child1> | ReturnType<typeof Child2>>
>
```

Or change your `InferError` (and set a constraint on the generic to have nice reporting in case you pass something else):

```ts
type InferError<T extends (...args: any[]) => Result<any, any>> = T extends (
  ...args: any[]
) => Result<any, infer E>
  ? E
  : never;
```

```ts
function Parent(): Result<null, InferError<typeof Child1 | typeof Child2>>
```

Full example: https://stackblitz.com/edit/m21xpf-clasia?file=index.ts

**Reply by pand-app** [reply]
Oh. Thanks, I've missed that.
